### DISCUSSION TITLE
从1.0升级到2.0的一些问题

### DISCUSSION THREAD

**Discussion by stwflyfox** [opentiny/tiny-engine#959] [Q&A]
我在1.0时修改源码实现的功能，在2.0新架构中要如何实现:
1. 页面属性中的路由改为下拉选项:
![image](https://github.com/user-attachments/assets/86754f33-3cd9-4bc2-aff6-71db266b6bb4)
2. 在属性设置中自建的属性录入部件:
![image](https://github.com/user-attachments/assets/98b4af2a-be03-4480-90e1-af78d8b30f13)
![image](https://github.com/user-attachments/assets/44833fbd-086e-41fe-90ff-a62b45d09abe)
3.在高级配置里的显示修改：
![image](https://github.com/user-attachments/assets/b66807e7-1041-42e6-a427-9b3074f8b30a)

**Comment by hexqi** [comment]
在新的架构下主要通过配置项、组件/模块替换、部分注册API实现功能的扩展和替换，核心是registry.js文件，对于这几种情况，可以这样处理：
(前置：先使用cli创建了全新的设计器项目)
1. 对UI做了调整扩展
解决方法：覆盖Page模块的基础设置部分来实现
![image](https://github.com/user-attachments/assets/36b6f6a7-5151-41c7-bfa9-db5cca22dc36)

```diff
// registry.js
export default {
   plugins: [
--     Page
++     {
++        ...Page,
++       components: {
++            ...Page.components,
++            PageGeneral:  YourCustomPageGeneral  // 实现自定义的基础设置组件来进行覆盖，上图红框部分对应该组件
++       }
++     }
   ]，
   // ...
}

```

2. 自定义的设置器
解决方法：注册设置器组件并在物料文件中使用
a) 注册
开发并导出自己的设置器组件，例如YourFieldConfigurator（configurators会在tiny-engine init时自动注册），开发configurator组件参考之前的Meta组件
```js
// src/configurators/index.js
import YourFieldConfigurator from './YourFieldConfigurator.vue' // 自定义的设置器组件

export const configurators = {
  YourFieldConfigurator, 
}
```

b）配置到物料文件bundle.json中
使用YourFieldConfigurator 自定义设置器组件
```json
"widget": {
  "component": "YourFieldConfigurator", 
  "props": {}
}
```

3.没太看明白意思，是否是期望：页面元素显示隐藏是通过接口查询得到的
如果是这种场景，似乎跟1.0/2.0版本没有关系，通过将“是否显示“绑定到状态变量上再通过调用接口查询后更新状态变量的方式，是否能够满足诉求？
![image](https://github.com/user-attachments/assets/b10c53d4-b428-4c54-8e98-e89ce1d02a91)

**Reply by stwflyfox** [reply]
非常感谢hexqi老师的回复，关于第3个问题是想把显示隐藏的选择方式，改成待合自己业务需求的代码
![image](https://github.com/user-attachments/assets/1de344ec-29b6-4ae1-9a49-8a2d90a7bb7a)

**Reply by stwflyfox** [reply]
想在是否显示的现有两种选择方式后面， 再增加一种我们自己的选择方式。

**Reply by hexqi** [reply]
这种当前只能替换整个高级属性配置模块来实现，暂时没有细粒度的配置或者覆盖方法。
类似第一个问题中的方式，例如，替换Events插件为重新开发的YourEventsPlugin，然后修改registry.js文件：
```js
// registry.js
export default {
   settings: [
     // ...
--   Events
++   YourEventsPlugin
   ]，
   // ...
}
```

**Comment by stwflyfox** [comment]
谢谢，我先尝试修改看看，有问题再请教。
